src/celeritas/grid/XsCalculator.hh

```diff
@@ -90,107 +84,67 @@ class XsCalculator
     {
-        // Interpolate *linearly* on energy
-        real_type upper_xs = this->get(lower_idx + 1);
-        if (lower_idx + 1 == data_.prime_index)
-        {
-            // Cross section data for the upper point is scaled by E: calculate
-            // the unscaled value
-            upper_xs /= upper_energy;
-        }
-
-        result = LinearInterpolator<real_type>(
-            {lower_energy, this->get(lower_idx)},
-            {upper_energy, upper_xs})(value_as<Energy>(energy));
-
-        if (lower_idx >= data_.prime_index)
-        {
-            result /= energy.value();
-        }
+        // Linear or cubic spline interpolation
+        result = UniformLogGridCalculator(grid, reals_)(energy);
     }
     else
     {
-        // Use cubic spline interpolation
-        CELER_ASSERT(data_.prime_index == XsGridData::no_scaling());
-        CELER_ASSERT(lower_idx + 1 < data_.derivative.size());
-        real_type lower_deriv = reals_[data_.derivative[lower_idx]];
-        real_type upper_deriv = reals_[data_.derivative[lower_idx + 1]];
-
-        result = SplineInterpolator<real_type>(
-            {lower_energy, (*this)[lower_idx], lower_deriv},
-            {upper_energy, (*this)[lower_idx + 1], upper_deriv})(
-            value_as<Energy>(energy));
+        // Spline interpolation without continuous derivatives
+        result = SplineCalculator(grid, reals_)(energy);
+    }
+    if (use_scaled)
+    {
+        result /= value_as<Energy>(energy);
     }
     return result;
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Get the cross section at the given index.
+ * Get the minimum energy.
  */
-CELER_FUNCTION real_type XsCalculator::operator[](size_type index) const
+CELER_FUNCTION auto XsCalculator::energy_min() const -> Energy
 {
-    real_type result = this->get(index);
-    if (index >= data_.prime_index)
-    {
-        result /= std::exp(loge_grid_[index]);
-    }
-    return result;
+    return Energy(std::exp(data_.lower ? data_.lower.grid.front
+                                       : data_.upper.grid.front));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the maximum energy.
+ */
+CELER_FUNCTION auto XsCalculator::energy_max() const -> Energy
+{
+    return Energy(
+        std::exp(data_.upper ? data_.upper.grid.back : data_.lower.grid.back));
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Get the raw cross section data at a particular index.
+ * Whether to use the scaled cross section grid.
  */
-CELER_FUNCTION real_type XsCalculator::get(size_type index) const
+CELER_FUNCTION bool XsCalculator::use_scaled(Energy energy) const
 {
-    CELER_EXPECT(index < data_.value.size());
-    return reals_[data_.value[index]];
+    real_type loge = std::log(value_as<Energy>(energy));
+    return !data_.lower || (data_.upper && loge >= data_.upper.grid.front);
```

Comment:
It shouldn't be possible to have neither `lower` nor `upper`, and it might be good to defer the call to std::log.
```suggestion
    return data_.upper && std::log(value_as<Energy>(energy)) >= data_.upper.grid.front;
```

Reply:
True, thanks!

Reply:
Actually I think it was ok before, otherwise we're not  correctly handling the case where there are only scaled values and the xs is below the grid.